OOP school library: set up associations

khanh-devos opened the pull request:
- [x] Create a class Classroom with the following:
@label instance variable, should be initialized in the constructor.
Setter and getter for @label (remember about attr_accessor).
Create the has-many/belongs-to relationship between Classroom and Student. The following should be implemented:
Create the has-many side (a classroom has many students).
Create the belongs-to side (a student belongs to a classroom).
Make sure that when adding a student to a classroom it also sets the classroom for the student.
Make sure that when setting the classroom for a student it also adds it to the classrooms' students.

- [x] Create a class Book with the following:
@title and @author instance variables, should be initialized in the constructor.
Setters and getters for instance variables (remember about attr_accessor).

- [x] Create a class Rental with the following:
@date instance variable, should be initialized in the constructor.
Setter and getter for @date (remember about attr_accessor).
Create the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. The following should be implemented:
Create the has-many side of Book and Rental (a book has many rentals).
Create the belongs-to side of Rental and Book (a rental belongs to a book).
Create the has-many side of Person and Rental (a person has many rentals).
Create the belongs-to side of Rental and Person (a rental belongs to a person).
Modify the constructor of Rental so Book and Person are set in it.

khanh-devos committed:
Build an association for classes: Book, Person, Rental, Classroom, ..
- Modify classes: Person, Student, Teacher
- Created new classes: Book, Classroom, Rental
- Created Associations:
1. Classroom : one-to-many : Student
2. Book: one-to-many : Rental
3. Person: one-to-many : Rental

khanh-devos committed:
fix rubocop errors

khanh-devos committed:
fix rubocop errors

Gambit142 commented on composition/rental.rb:
```diff
@@ -0,0 +1,29 @@
+    @person = person
+    person.rentals << self
+  end
+
+  def book=(book)
+    @book = book
+    book.rentals.push(self) unless book.rentals.include?(self)
+  end
+
+  def person=(person)
+    @person = person
+    person.rentals.push(self) unless person.rentals.include?(self)
+  end
```
- [x] Great work creating these codes, I can see that you are manually creating the _**setter**_ methods instead of using the `attr_accessor` method. Kindly remove this line of codes from each setter method because you have them written in the `initialize` method of this class:

```ruby
 person.rentals.push(self) unless person.rentals.include?(self)
```
```ruby
 book.rentals.push(self) unless book.rentals.include?(self)
```

Gambit142 commented on composition/book.rb:
```diff
@@ -0,0 +1,12 @@
+    @author = author
+    @rentals = []
+  end
+end
```
- [x] The `Book` and the `Person` class have a many-to-many relationship through the `Rental` class. Therefore, each class instance needs to keep track of its rentals. Kindly implement this in this class. 👍 💻 

#### Hint:

```ruby
def add_rental(person, date)
  Rental.new(date, self, person)
end
```

Gambit142 commented on composition/book.rb:
```diff
@@ -0,0 +1,12 @@
+require 'securerandom'
+
+class Book
+  attr_reader :id, :rentals
```
- [ ] Based on the requirement, kindly create a setter method for the rentals attribute. You can use this shorthand method 🖱️ 💻:

```ruby
attr_accessor :rentals
```

Gambit142 commented on person.rb:
```diff
@@ -1,16 +1,17 @@
-require './nameable'
+require './decorators/nameable'
 require 'securerandom'
 
 class Person < Nameable
-  attr_reader :id
-  attr_accessor :name, :age
```
- [x] The `Book` and the `Person` class have a many-to-many relationship through the `Rental` class. Therefore, each class instance needs to keep track of its rentals. Kindly implement this in this class. 👍 💻 

#### Hint:

```ruby
def add_rental(book, date)
  Rental.new(date, book, self)
end
```

Gambit142 commented on person.rb:
```diff
@@ -1,16 +1,17 @@
 class Person < Nameable
-  attr_reader :id
-  attr_accessor :name, :age
+  attr_reader :id, :rentals
+  attr_accessor :name, :age, :parent_permission
```
- [x] Based on the requirement, kindly create a setter method for the rentals attribute. You can use this shorthand method 🖱️ 💻:

```ruby
attr_accessor :rentals
```

Gambit142 reviewed (changes requested):
Hi @khanh-devos 👋, 

Good job so far!

## Highlights
- Setup linters correctly ✔️ 
- Created new classes based on requirements ✔️ 
- Setup association between classes correctly ✔️

You still need to work on some issues to go to the next project, but you are almost there!

## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend considering them as they can improve your code._

You can also consider the following:

N/A

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### Please tag me in your question so I can receive the notification. You can tag me by typing @Gambit142

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

khanh-devos committed:
Modify following the code review requirement
- Set attr_accessor for rentals of Book and Person
- Created method add_rental for Book and Person
- Set attr_accessor for attribute book and person

khanh-devos committed:
Fix rubocop style errors

roshan-bajgain reviewed (approved):
Hi @khanh-devos, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- No linter errors. ✔️
- Gitflow is followed. ✔️
- Good job on implementing the changes ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

khanh-devos commented:
Thank you very much @roshan-bajgain and @Gambit142
